### PR TITLE
Add support for isDriven in SyncPorts

### PR DESCRIPTION
**Pull request by sgundoji**

### CODE REVIEW

File: sparta/sparta/ports/SyncPort.hpp [hunk 1]
```diff
@@ -194,6 +194,37 @@ namespace sparta
             return sync_in_port_->getRawReady_();
         }
 
+        /*! \brief Determine if this DataOutPort has any connected
+         *        DataInPort where the data is to be delivered on the
+         *        given cycle.
```
Comment (knute-sifive) [inline sparta/sparta/ports/SyncPort.hpp:199]: Can you fix the comments to reflect this is a SyncPort?

File: sparta/sparta/ports/SyncPort.hpp [hunk 2]
```diff
@@ -194,6 +194,37 @@ namespace sparta
...
+            for(SyncInPort<DataT>* itr : bound_in_ports_) {
+                if(itr->isDriven(rel_cycle)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        //! \brief Does this DataOutPort have _any_ DataInPort's where
+        //!        the data is not yet delivered?
```
Comment (knute-sifive) [inline sparta/sparta/ports/SyncPort.hpp:216]: Comment fix...

File: sparta/sparta/ports/SyncPort.hpp [hunk 3]
```diff
@@ -158,7 +158,7 @@ namespace sparta
 
             OutPort::bind(in);
 
-            bound_ports_.push_back(in);
+            bound_in_ports_.push_back(inp);
```
Comment (knute-sifive) [inline sparta/sparta/ports/SyncPort.hpp:161]: Ugh, good catch -- old code was storing the in port in the same vector -- twice!  I know you needed a new vector...

File: sparta/sparta/ports/SyncPort.hpp [hunk 4]
```diff
@@ -438,6 +472,28 @@ namespace sparta
...
         using Port::bind;
 
+        /*!
+         * \brief Determine if this DataInPort is driven on the
+         *        given cycle
+         * \param rel_cycle The relative cycle (from now) the data
+         *                  will be delivered
+         * \return true if driven at the given cycle (data not yet delivered)
+         * \note If the DataInPort was driven with a zero-cycle delay,
+         *       this function will always return false.
```
Comment (knute-sifive) [inline sparta/sparta/ports/SyncPort.hpp:482]: Fix comments.

File: sparta/sparta/ports/SyncPort.hpp [hunk 5]
```diff
@@ -438,6 +472,28 @@ namespace sparta
...
+         */
+        bool isDriven(Clock::Cycle rel_cycle) const override {
+            return forward_event_->isScheduled(rel_cycle);
+        }
+
+        /*! \brief Is this Port driven at all?
+         *  \return true if driven at all (data not yet delivered)
+         *  \note If the DataInPort was driven with a zero-cycle delay,
+         *       this function will always return false.
```
Comment (knute-sifive) [inline sparta/sparta/ports/SyncPort.hpp:491]: Fix comments.

**Review by knute-sifive — DISMISSED** [review]
When you update the comments, I'll merge this in.  Thanks for doing this!

File: sparta/sparta/ports/SyncPort.hpp (hunk 1 above)
Comment (sgundoji) [reply sparta/sparta/ports/SyncPort.hpp]: All comments fixed in next commit.

**Comment by knute-sifive** [comment]
Talked with Spurthi -- he will be adding testing and a fix for `isDriven` that he discovered.

File: sparta/test/SyncPort/SyncPort_test.cpp [hunk 6]
```diff
@@ -425,6 +425,271 @@ TestSystem::~TestSystem()
...
+    public:
+        ParameterSet(sparta::TreeNode* tn) : sparta::ParameterSet(tn) {}
+    };
+
+    Destn(sparta::TreeNode* node, const Destn::ParameterSet*);
+    ~Destn();
+
+    // Callback for data.  This method is to test that the command is
+    // received before the data.  The data is ignored.
+    void data_callback(const char &);
```
Comment (knute-sifive) [inline sparta/test/SyncPort/SyncPort_test.cpp:548]: Nit: function naming convention does not match Sparta standards.

File: sparta/test/SyncPort/SyncPort_test.cpp [hunk 7]
```diff
@@ -425,6 +425,271 @@ TestSystem::~TestSystem()
...
+    sparta::SyncInPort<char>     in_data;
+
+    static constexpr const char* name = "Destn";
+
+private:
+    // Event set for the unit
+    sparta::EventSet ev_set;
+
+    // Self-scheduling event
+    sparta::UniqueEvent<> ev_do_work;
```
Comment (knute-sifive) [inline sparta/test/SyncPort/SyncPort_test.cpp:564]: Nit: member naming convention does not match Sparta standards (missing trailing `_`)

File: sparta/test/SyncPort/SyncPort_test.cpp [hunk 8]
```diff
@@ -425,6 +425,271 @@ TestSystem::~TestSystem()
...
+        ParameterSet(sparta::TreeNode* tn) : sparta::ParameterSet(tn) {}
+    };
+
+    Source(sparta::TreeNode* node, const Source::ParameterSet*);
+    ~Source();
+
+    // Called before simulation by the testing framework to
+    //    - Send commands on the SyncOut port to the destn
+    //    - Calculate the expected input Ticks and Data from the destn
+    void schedule_commands();
```
Comment (knute-sifive) [inline sparta/test/SyncPort/SyncPort_test.cpp:447]: Nit: method naming convention does not match Sparta standards

**Review by knute-sifive — APPROVED** [review]

**Pull request merged**